### PR TITLE
Shortest-path routing for Spider, end-to-end testing framework with adjustable delay

### PR DESCRIPTION
**Pull request by yangl1996**
Todo: 
- [x] run all LND tests

Deferred:
- #6 
- a readme for how to run the spider test

### CODE REVIEW

File: routing/router.go [hunk 1]
```diff
@@ -1606,6 +1606,52 @@ func (r *ChannelRouter) SendToRoute(routes []*Route,
 	return r.sendPayment(payment, paySession)
 }
 
+// SendToShortestPath attempts to send a payment as described within the
+// passed LightningPayment through the shortest path returned by
+// paySession.RequestRoute. This function is blocking and will return 
+// either: when the payment is successful, or all routes
+// have been attempted and resulted in a failed payment. If the payment
```
Comment (vibhaa) [inline routing/router.go:1622]: Does this attempt all routes or just the shortest one? So even if that one fails it will return failed payment right?

File: routing/router_test.go [hunk 2]
```diff
@@ -920,6 +920,131 @@ func TestSendPaymentErrorPathPruning(t *testing.T) {
...
+
+	const startingBlockHeight = 101
+	ctx, cleanUp, err := createTestCtxFromFile(startingBlockHeight, basicGraphFilePath)
+	defer cleanUp()
+	if err != nil {
+		t.Fatalf("unable to create router: %v", err)
+	}
+
+	// Craft a LightningPayment struct that'll send a payment from roasbeef
+	// to luo ji for 1000 satoshis, with a maximum of 1000 satoshis in fees.
```
Comment (vibhaa) [inline routing/router_test.go]: it says "no fee limit" here though. Is there a fee then? Does the comment need some editing.

File: routing/router_test.go [hunk 3]
```diff
@@ -920,6 +920,131 @@ func TestSendPaymentErrorPathPruning(t *testing.T) {
...
+	}
+}
+
+// TestSendToShortestPathFailAtBrokenShortestPath checks that the payment will fail if
+// the shortest path is broken
+func TestSendToShortestPathFailAtBrokenShortestPath(t *testing.T) {
+	t.Parallel()
+
+	const startingBlockHeight = 101
+	ctx, cleanUp, err := createTestCtxFromFile(startingBlockHeight, basicGraphFilePath)
```
Comment (vibhaa) [inline routing/router_test.go:992]: What does the topology look like for this? Just for context?

File: lnrpc/rpc.proto [hunk 4]
```diff
@@ -717,6 +717,8 @@ message SendRequest {
     send the payment.
     */
     FeeLimit fee_limit = 8;
+
+    int32 spiderAlgo = 9;
```
Comment (vibhaa) [inline lnrpc/rpc.proto:725]: Can you please add a comment here like the above fields explaining what this parameter is?

File: lnd_test.go [hunk 5]
```diff
@@ -3644,6 +3645,704 @@ func testMultiHopPayments(net *lntest.NetworkHarness, t *harnessTest) {
 	closeChannelAndAssert(ctxt, t, net, carol, chanPointCarol, false)
 }
 
+func testSpiderShortestPath(net *lntest.NetworkHarness, t *harnessTest) {
```
Comment (vibhaa) [inline lnd_test.go:3652]: Can you add a comment here detailing what this test is about (similar to all the above tests)?

File: lnd_test.go [hunk 6]
```diff
@@ -3644,6 +3645,704 @@ func testMultiHopPayments(net *lntest.NetworkHarness, t *harnessTest) {
...
+
+	ctxt, _ = context.WithTimeout(ctxb, timeout)
+	chanPointn1n2 := openChannelAndAssert(
+		ctxt, t, net, n1, n2,
+		lntest.OpenChannelParams{
+			Amt: chanAmt,
+			PushAmt: pushAmt
+		},
+	)
+	networkChans = append(networkChans, chanPointn1n2)
```
Comment (vibhaa) [inline lnd_test.go]: Can you also add a comment for every block here as to what channel its opening and what hash its collecting? Just a line or two would be great for like n1n2 and then n2n3 and so on.

File: lnd_test.go [hunk 7]
```diff
@@ -3912,366 +3775,97 @@ func testSpiderShortestPath(net *lntest.NetworkHarness, t *harnessTest) {
...
-					resp.PaymentError)
-				return
+			if err := payStreams[payIntents[j][0]].Send(sendReq); err != nil {
+				t.Fatalf("unable to send payment: "+
+					"%v", err)
 			}
 		}
-		errChan <- nil
-	}()
+	}
```
Comment (vibhaa) [inline lnd_test.go]: Can you add a comment saying "receive responses" or something brief?

File: lnd_test.go [hunk 8]
```diff
@@ -3659,8 +3659,12 @@ func testSpiderShortestPath(net *lntest.NetworkHarness, t *harnessTest) {
 		return channelInfo.Channels, nil
 	}*/
 	
+	// Set channel sizes here. chanAmt sets the channel
 	const chanAmt = btcutil.Amount(200000)
 	const pushAmt = btcutil.Amount(100000)
+
+	baseFee := int64(0)
+	feeRate := int64(1)
```
Comment (vibhaa) [inline lnd_test.go]: can you add a comment explaining these two fields? In particular, what is feeRate?

File: lnd_test.go [hunk 9]
```diff
@@ -3801,69 +3778,110 @@ func testSpiderShortestPath(net *lntest.NetworkHarness, t *harnessTest) {
...
+	timeLockDelta := uint32(144)
+
+	expectedPolicy := &lnrpc.RoutingPolicy{
+		FeeBaseMsat:      baseFee,
+		FeeRateMilliMsat: feeRate * testFeeBase,
+		TimeLockDelta:    timeLockDelta,
+		MinHtlc:          1000, // default value
+	}
 
+	graphSubs := make([]graphSubscription, numNodes)
```
Comment (vibhaa) [inline lnd_test.go]: What is this graphSubscription about?

Comment (vibhaa) [inline lnd_test.go:3831]: Can you add some more comments on what exactly is happening here? Did you add these graphNotifications? If so, what are they/where are they? If not, can you expand on how you're using them?

File: lnd_test.go [hunk 10]
```diff
@@ -3801,69 +3778,110 @@ func testSpiderShortestPath(net *lntest.NetworkHarness, t *harnessTest) {
...
-			if err := payStreams[payIntents[j][0]].Send(sendReq); err != nil {
-				t.Fatalf("unable to send payment: "+
-					"%v", err)
-			}
+	expectedMessages := make([]expectedChanUpdate, numChannels)
+	for i := 0; i < numChannels; i++ {
+		expectedMessages[i] = expectedChanUpdate{
+			nodes[connections[i][0]].PubKeyStr, expectedPolicy, chanPoints[i]}
+	} 
+
```
Comment (vibhaa) [inline lnd_test.go]: Can you add a brief comment on what exactly is happening here? In particular, why do we need to "update" the policy?

File: lnd_test.go [hunk 11]
```diff
@@ -3801,69 +3778,110 @@ func testSpiderShortestPath(net *lntest.NetworkHarness, t *harnessTest) {
...
-			if err != nil {
-				t.Fatalf(err.Error())
-			}
-		case <-time.After(maxTime):
-			t.Fatalf("waiting for payments to finish too long "+
-				"(%v)", maxTime)
-		}
-	}
+	wg.Wait()
+
```
Comment (vibhaa) [inline lnd_test.go]: Similarly, can you add a comment saying "Collecting statistics - success rate and volume or something like that?"

File: lnd_test.go [hunk 12]
```diff
@@ -3826,6 +3826,7 @@ func testSpiderShortestPath(net *lntest.NetworkHarness, t *harnessTest) {
 	// Initialize seed random in order to generate invoices.
 	prand.Seed(time.Now().UnixNano())
 	var wg sync.WaitGroup
+	var tranwg sync.WaitGroup
```
Comment (vibhaa) [inline lnd_test.go:3887]: Can you add a comment to explain what these wait groups are doing/how they are organized (like per demand pair or intent) and how they are relevant below?

File: routing/missioncontrol.go [hunk 13]
```diff
@@ -361,7 +361,7 @@ func (p *paymentSession) RequestRoute(payment *LightningPayment,
 	// Otherwise we actually need to perform path finding, so we'll obtain
 	// our current prune view snapshot. This view will only ever grow
 	// during the duration of this payment session, never shrinking.
-	pruneView := p.pruneViewSnapshot
+	pruneView := graphPruneView{}
```
Comment (vibhaa) [inline routing/missioncontrol.go]: What exactly is this change doing? Just curious

File: lnd_test.go [hunk 14]
```diff
@@ -3796,10 +3797,12 @@ func testSpiderShortestPath(net *lntest.NetworkHarness, t *harnessTest) {
 		defer close(graphSubs[i].quit)
 	}
 
-	expectedMessages := make([]expectedChanUpdate, numChannels)
+	expectedMessages := make([]expectedChanUpdate, numChannels * 2)
 	for i := 0; i < numChannels; i++ {
-		expectedMessages[i] = expectedChanUpdate{
+		expectedMessages[i * 2] = expectedChanUpdate{
```
Comment (vibhaa) [inline lnd_test.go]: Same question: What exactly is happening here?

File: chainregistry.go [hunk 15]
```diff
@@ -33,14 +33,14 @@ import (
 )
 
 const (
-	defaultBitcoinMinHTLCMSat   = lnwire.MilliSatoshi(1000)
-	defaultBitcoinBaseFeeMSat   = lnwire.MilliSatoshi(1000)
-	defaultBitcoinFeeRate       = lnwire.MilliSatoshi(1)
+	defaultBitcoinMinHTLCMSat   = lnwire.MilliSatoshi(0)
```
Comment (vibhaa) [inline chainregistry.go:36]: This is okay for now, but can we add an issue to track this? We might want to change it back at some point (so as to not affect LND)

File: lnd_test.go [hunk 16]
```diff
@@ -3873,69 +3873,42 @@ func testSpiderShortestPath(net *lntest.NetworkHarness, t *harnessTest) {
...
-		t.Fatalf("bob's local balance is incorrect, got %v, expected"+
-			" %v", bobInfo.LocalBalance, bobAmt)
-	}
-	if bobInfo.RemoteBalance != aliceAmt {
-		t.Fatalf("bob's remote balance is incorrect, got %v, "+
-			"expected %v", bobInfo.RemoteBalance, aliceAmt)
+	succeeded := make([]uint64, numPayIntents)
+	failed := make([]uint64, numPayIntents)
+	var totSucceeded uint64 = 0
+	var totFailed uint64 = 0
```
Comment (vibhaa) [inline lnd_test.go]: Please add another comment here saying "aggregating per intent or demand pair success rates.

File: lnrpc/rpc.swagger.json [hunk 17]
```diff
@@ -2634,6 +2634,10 @@
         "fee_limit": {
           "$ref": "#/definitions/lnrpcFeeLimit",
           "description": "*\nThe maximum number of satoshis that will be paid as a fee of the payment.\nThis value can be represented either as a percentage of the amount being\nsent, or as a fixed amount of the maximum fee the user is willing the pay to\nsend the payment."
+        },
+        "spiderAlgo": {
+          "type": "integer",
+          "format": "int32"
```
Comment (vibhaa) [inline lnrpc/rpc.swagger.json]: I would add a description here too, to clarify what this is about.

File: lntest/harness.go [hunk 18]
```diff
@@ -239,8 +239,13 @@ func (n *NetworkHarness) TearDownAll() error {
 // NewNode fully initializes a returns a new HarnessNode bound to the
 // current instance of the network harness. The created node is running, but
 // not yet connected to other nodes within the network.
-func (n *NetworkHarness) NewNode(name string, extraArgs []string) (*HarnessNode, error) {
-	return n.newNode(name, extraArgs, false)
+func (n *NetworkHarness) NewNode(name string, extraArgs []string, delay ...int) (*HarnessNode, error) {
```
Comment (vibhaa) [inline lntest/harness.go]: Can you add the delay explanation to the comments also?

File: lntest/node.go [hunk 19]
```diff
@@ -111,6 +111,8 @@ type nodeConfig struct {
 	P2PPort  int
 	RPCPort  int
 	RESTPort int
+
+	Delay int
```
Comment (vibhaa) [inline lntest/node.go]: Can you add a comment explaining what this delay is?

File: routing/missioncontrol.go [hunk 20]
```diff
@@ -331,6 +331,34 @@ func (p *paymentSession) ReportChannelFailure(e uint64) {
 	p.mc.Unlock()
 }
 
+func (p *paymentSession) RequestShortestPath(payment *LightningPayment,
+	height uint32, finalCltvDelta uint16) (*Route, error) {
+
+	// Taking into account this prune view, we'll attempt to locate a path
```
Comment (vibhaa) [inline routing/missioncontrol.go]: "locate the shortest path"

File: lnd_test.go [hunk 21]
```diff
@@ -3644,6 +3645,462 @@ func testMultiHopPayments(net *lntest.NetworkHarness, t *harnessTest) {
...
+	}
+
+	// Channels to signal the end of the experiment
+	var paymentStop = make([]chan int, numPayIntents)
+	for i := range paymentTick {
+   		paymentStop[i] = make(chan int, 1)
+	}
+	
+	// Dispatcher goroutines
+	for i := 0; i < numPayIntents; i++ {
```
Comment (vibhaa) [inline lnd_test.go:3990]: Any chance you can refactor this code and give all the goroutines names? Its just really hard to follow the whole 3 layers of goroutines. I get the flow - but might help if there were some descriptive names/self contained functions.

File: lnrpc/rpc.pb.go [hunk 22]
```diff
@@ -640,7 +640,8 @@ type SendRequest struct {
 	// This value can be represented either as a percentage of the amount being
 	// sent, or as a fixed amount of the maximum fee the user is willing the pay to
 	// send the payment.
-	FeeLimit *FeeLimit `protobuf:"bytes,8,opt,name=fee_limit,json=feeLimit" json:"fee_limit,omitempty"`
+	FeeLimit   *FeeLimit `protobuf:"bytes,8,opt,name=fee_limit,json=feeLimit" json:"fee_limit,omitempty"`
+	SpiderAlgo int32     `protobuf:"varint,9,opt,name=spiderAlgo" json:"spiderAlgo,omitempty"`
```
Comment (vibhaa) [inline lnrpc/rpc.pb.go]: Please add a comment

File: lnrpc/rpc.pb.go [hunk 23]
```diff
@@ -704,6 +705,13 @@ func (m *SendRequest) GetFeeLimit() *FeeLimit {
 	return nil
 }
 
+func (m *SendRequest) GetSpiderAlgo() int32 {
```
Comment (vibhaa) [inline lnrpc/rpc.pb.go:711]: Please add a single line comment atleast here

File: lntest/node.go [hunk 24]
```diff
@@ -331,6 +333,13 @@ func (hn *HarnessNode) start(lndError chan<- error) error {
 		hn.logFile = file
 	}
 
+	// we first apply delay config to the node
```
Comment (vibhaa) [inline lntest/node.go]: maybe mention using "tcconfig we create a new delay shell" or whatever the right phrase is?

File: routing/missioncontrol.go [hunk 25]
```diff
@@ -331,6 +331,34 @@ func (p *paymentSession) ReportChannelFailure(e uint64) {
 	p.mc.Unlock()
 }
 
+func (p *paymentSession) RequestShortestPath(payment *LightningPayment,
```
Comment (vibhaa) [inline routing/missioncontrol.go:337]: Comment for the function please?

File: routing/router.go [hunk 26]
```diff
@@ -36,6 +36,11 @@ const (
 	defaultPayAttemptTimeout = time.Duration(time.Second * 60)
 )
 
+const (
```
Comment (vibhaa) [inline routing/router.go:39]: What are these constants for? Can we add some comments?

File: rpcserver.go [hunk 27]
```diff
@@ -1971,6 +1971,8 @@ type rpcPaymentIntent struct {
 	routeHints [][]routing.HopHint
 
 	routes []*routing.Route
+
+	spiderAlgo int
```
Comment (vibhaa) [inline rpcserver.go:1978]: Comment will be helpful

**Comment by vibhaa** [comment]
Most of these are minor nits/addition on comments. There's only one place in the testing code where I felt a little code refactoring could be helpful. Otherwise looks good

**Review by vibhaa — CHANGES_REQUESTED** [review]
I think most of these are fine except for needing a few more comments. My only issue was with some code in the tests that I felt could be refactored a little. Other than that would recommend fixing the 3 things you yourself mentioned also. (Readme/removing it from itest would be very helpful) and please do run make itest before merging.

**Comment by yangl1996** [comment]
Thanks for the review! I've added most of the requested comments. I'll refactor the `lnd_test` test case  later today, and fix the itest tomorrow - it currently fails, maybe because I changed the minHTLC and minFees values.

BTW, I marked requests that I have fixed as resolved, so it is easier to track which requests have not been fixed. Feel free to reopen :)
